### PR TITLE
Nvfuser python api patch take 2

### PR DESCRIPTION
**Pull request by jjsjann123**
landing #83645 again.

Previously we are breaking on codegen bf16 kernel for cuda TK 10.2. Added a short-cut to disable bf tests on pre cuda 11 build.

### CODE REVIEW

**Comment by jjsjann123** [comment]
Added trunk here, since I"m not seeing cuda10.2 in CI runs yet.

**Comment by ngimel** [comment]
As with TS integration, can we make it so that nvfuser errors out earlier and with nicer error message when bfloat is encountered? (Btw, bfloat also won't work on pre-ampere devices, and I'm not sure python integration stack currently takes that into account)

**Comment by jjsjann123** [comment]
> As with TS integration, can we make it so that nvfuser errors out earlier and with nicer error message when bfloat is encountered? (Btw, bfloat also won't work on pre-ampere devices, and I'm not sure python integration stack currently takes that into account)

That is our goal as well, but that part likely would happen on a separate PR. We can do some assert logic when parsing inputs on cpp side and assert on no bf16 appearing for build with earlier cuda TK, that should give us better error message, other than invalid kernel code.

Regarding bfloat16 working on pre-ampere devices: I'm still trying to figure out what's going on there. We only takes bfoat16 at Input/output tensors and are NOT using any bf intrinsics.
Per PTX ISA doc, it seems to suggest that the inlined PTX instruction that we used for casting to bf16 requires sm_80+. But apparently the code runs fine on volta. I'm not sure if I'm misreading the doc or if there's some inconsistency there. I'm following up with the compiler team and hopefully would get a better answer from them.

**Comment by ngimel** [comment]
I'm not suggesting it should be done in this PR. I don't think the code runs on volta (`cvt.rn.bf16.f32` is sm80+ only, as per ptx manual), here you probably got lucky in only needing to convert bfloat16 to float (which indeed works for earlier devices), and not the other way around.

**Comment by jjsjann123** [comment]
This PR looks good this time. All tests including the ones in cuda 10.2 is passing. I can merge once I get a stamp. :bow:

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -180,11 +180,12 @@ void initNvFuserPythonBindings(PyObject* module) {
       .def(
           "define_constant",
           [](nvfuser::FusionDefinition& self,
-             c10::complex<double> val) -> nvfuser::Scalar* {
+             std::complex<double> val) -> nvfuser::Scalar* {
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:183]: Did you mean for this to be `std::complex` instead of `c10::complex`?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:183]: Yes. With std::complex<double>, python complex doesn't resolve to this overload, but went for the other function with `bool val` instead. And that's the root cause of the test failure earlier.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**